File: public/js/codenames.js
let socket = io({path: window.location.pathname + 'socket.io'}) // Connect to server


// Sign In Page Elements
////////////////////////////////////////////////////////////////////////////
// Divs
let joinDiv = document.getElementById('join-game')
let joinErrorMessage = document.getElementById('error-message')
// Input Fields
let joinNickname = document.getElementById('join-nickname')
let joinRoom = document.getElementById('join-room')
let joinPassword = document.getElementById('join-password')
// Buttons
let joinEnter = document.getElementById('join-enter')
let joinCreate = document.getElementById('join-create')


// Game Page Elements
////////////////////////////////////////////////////////////////////////////
// Divs
let gameDiv = document.getElementById('game')
let clueEntryDiv = document.getElementById('clue-form')
let boardDiv = document.getElementById('board')
let aboutWindow = document.getElementById('about-window')
let afkWindow = document.getElementById('afk-window')
let serverMessageWindow = document.getElementById('server-message')
let serverMessage = document.getElementById('message')
let overlay = document.getElementById('overlay')
let logDiv = document.getElementById('log')
// Buttons
let leaveRoom = document.getElementById('leave-room')
let joinRed = document.getElementById('join-red')
let joinBlue = document.getElementById('join-blue')
let randomizeTeams = document.getElementById('randomize-teams')
let endTurn = document.getElementById('end-turn')
let newGame = document.getElementById('new-game')
let clueDeclareButton = document.getElementById('declare-clue')
let buttonRoleGuesser = document.getElementById('role-guesser')
let buttonRoleSpymaster = document.getElementById('role-spymaster')
let toggleDifficulty = document.getElementById('player-difficulty')
let buttonDifficultyNormal = document.getElementById('difficulty-normal')
let buttonDifficultyHard = document.getElementById('difficulty-hard')
let buttonModeCasual = document.getElementById('mode-casual')
let buttonModeTimed = document.getElementById('mode-timed')
let buttonConsensusSingle = document.getElementById('consensus-single')
let buttonConsensusConsensus = document.getElementById('consensus-consensus')
let buttonAbout = document.getElementById('about-button')
let buttonAfk = document.getElementById('not-afk')
let buttonServerMessageOkay = document.getElementById('server-message-okay')
let buttonBasecards = document.getElementById('base-pack')
let buttonDuetcards = document.getElementById('duet-pack')
let buttonUndercovercards = document.getElementById('undercover-pack')
let buttonCustomcards = document.getElementById('custom-pack')
let buttonNsfwcards = document.getElementById('nsfw-pack')// Clue entry
let clueWord = document.getElementById('clue-word')
let clueCount = document.getElementById('clue-count')
// Slider
let timerSlider = document.getElementById('timer-slider')
let timerSliderLabel = document.getElementById('timer-slider-label')
// Player Lists
let undefinedList = document.getElementById('undefined-list')
let redTeam = document.getElementById('red-team')
let blueTeam = document.getElementById('blue-team')
// UI Elements
let scoreRed = document.getElementById('score-red')
let scoreBlue = document.getElementById('score-blue')
let turnMessage = document.getElementById('status')
let timer = document.getElementById('timer')
let clueDisplay = document.getElementById('clue-display')


// init
////////////////////////////////////////////////////////////////////////////
// Default game settings
let playerRole = 'guesser'
let difficulty = 'normal'
let mode = 'casual'
let consensus = 'single'

// Show the proper toggle options
buttonModeCasual.disabled = true;
buttonModeTimed.disabled = false;
buttonRoleGuesser.disabled = true;
buttonRoleSpymaster.disabled = false;
buttonConsensusSingle.disabled = true;
buttonConsensusConsensus.disabled = false;

// Autofill room and password from query fragment
joinRoom.value = extractFromFragment(window.location.hash, 'room');
joinPassword.value = extractFromFragment(window.location.hash, 'password');


// UI Interaction with server
////////////////////////////////////////////////////////////////////////////
// User Joins Room
joinEnter.onclick = () => {       
  socket.emit('joinRoom', {
    nickname:joinNickname.value,
    room:joinRoom.value,
    password:joinPassword.value
  })
}
// User Creates Room
joinCreate.onclick = () => {      
  socket.emit('createRoom', {
    nickname:joinNickname.value,
    room:joinRoom.value,
    password:joinPassword.value
  })
}
// User Leaves Room
leaveRoom.onclick = () => {       
  socket.emit('leaveRoom', {})
}
// User Joins Red Team
joinRed.onclick = () => {         
  socket.emit('joinTeam', {
    team:'red'
  })
}
// User Joins Blue Team
joinBlue.onclick = () => {        
  socket.emit('joinTeam', {
    team:'blue'
  })
}
// User Randomizes Team
randomizeTeams.onclick = () => {  
  socket.emit('randomizeTeams', {})
}
// User Starts New Game
newGame.onclick = () => {         
  socket.emit('newGame', {})
}
clueDeclareButton.onclick = () => {
  socket.emit('declareClue', {word: clueWord.value, count: clueCount.value})
  clueWord.value = ''
  clueCount.value = 1
  return false
}
// User Picks spymaster Role
buttonRoleSpymaster.onclick = () => { 
  socket.emit('switchRole', {role:'spymaster'})
}
// User Picks guesser Role
buttonRoleGuesser.onclick = () => {   
  socket.emit('switchRole', {role:'guesser'})
}
// User Picks Hard Difficulty
buttonDifficultyHard.onclick = () => {
  socket.emit('switchDifficulty', {difficulty:'hard'})
}
// User Picks Normal Difficulty 
buttonDifficultyNormal.onclick = () => {
  socket.emit('switchDifficulty', {difficulty:'normal'})
}
// User Picks Timed Mode
buttonModeTimed.onclick = () => { 
  socket.emit('switchMode', {mode:'timed'})
}
// User Picks Casual Mode
buttonModeCasual.onclick = () => {
  socket.emit('switchMode', {mode:'casual'})
}
// User Picks Single Consensus Mode
buttonConsensusSingle.onclick = () => {
  socket.emit('switchConsensus', {consensus:'single'})
}
// User Picks Consensus Consensus Mode
buttonConsensusConsensus.onclick = () => {
  socket.emit('switchConsensus', {consensus:'consensus'})
}
// User Ends Turn
endTurn.onclick = () => {
  socket.emit('endTurn', {})
}
// User Clicks Tile
function tileClicked(i,j){
  socket.emit('clickTile', {i:i, j:j})
}
// User Clicks About
buttonAbout.onclick = () => {
  if (aboutWindow.style.display === 'none') {
    aboutWindow.style.display = 'block'
    overlay.style.display = 'block'
    buttonAbout.className = 'open above'
  } else {
    aboutWindow.style.display = 'none'
    overlay.style.display = 'none'
    buttonAbout.className = 'above'
  }
}
// User Clicks card pack
buttonBasecards.onclick = () => {
  socket.emit('changeCards', {pack:'base'})
}
// User Clicks card pack
buttonDuetcards.onclick = () => {
  socket.emit('changeCards', {pack:'duet'})
}
// User Clicks card pack
buttonUndercovercards.onclick = () => {
  socket.emit('changeCards', {pack:'undercover'})
}
/// User Clicks card pack
buttonCustomcards.onclick = () => {
  socket.emit('changeCards', {pack:'custom'})
}
// User Clicks card pack
buttonNsfwcards.onclick = () => {
  socket.emit('changeCards', {pack:'nsfw'})
}
// When the slider is changed
timerSlider.addEventListener("input", () =>{
  socket.emit('timerSlider', {value:timerSlider.value})
})

// User confirms theyre not afk
buttonAfk.onclick = () => {
  socket.emit('active')
  afkWindow.style.display = 'none'
  overlay.style.display = 'none'
}

// User confirms server message
buttonServerMessageOkay.onclick = () => {
  serverMessageWindow.style.display = 'none'
  overlay.style.display = 'none'
}

// Server Responses to this client
////////////////////////////////////////////////////////////////////////////
socket.on('serverStats', (data) => {        // Client gets server stats
  document.getElementById('server-stats').innerHTML = "Players: " + data.players + " | Rooms: " + data.rooms
})

socket.on('joinResponse', (data) =>{        // Response to joining room
  if(data.success){
    joinDiv.style.display = 'none'
    gameDiv.style.display = 'block'
    joinErrorMessage.innerText = ''
    updateFragment();
  } else joinErrorMessage.innerText = data.msg
})

socket.on('createResponse', (data) =>{      // Response to creating room
  if(data.success){
    joinDiv.style.display = 'none'
    gameDiv.style.display = 'block'
    joinErrorMessage.innerText = ''
    updateFragment();
  } else joinErrorMessage.innerText = data.msg
})

socket.on('leaveResponse', (data) =>{       // Response to leaving room
  if(data.success){
    joinDiv.style.display = 'block'
    gameDiv.style.display = 'none'
    wipeBoard();
  }
})

socket.on('timerUpdate', (data) => {        // Server update client timer
  timer.innerHTML = "[" + data.timer + "]"
})

socket.on('newGameResponse', (data) => {    // Response to New Game
  if (data.success){
    wipeBoard();
  }
})

socket.on('afkWarning', () => {    // Response to Afk Warning
  afkWindow.style.display = 'block'
  overlay.style.display = 'block'
})

socket.on('afkKicked', () => {    // Response to Afk Kick
  afkWindow.style.display = 'none'
  serverMessageWindow.style.display = 'block'
  serverMessage.innerHTML = 'You were kicked for being AFK'
  overlay.style.display = 'block'
})

socket.on('serverMessage', (data) => {    // Response to Server message
  serverMessage.innerHTML = data.msg
  serverMessageWindow.style.display = 'block'
  overlay.style.display = 'block'
})

socket.on('switchRoleResponse', (data) =>{  // Response to Switching Role
  if(data.success){
    playerRole = data.role;
    if (playerRole === 'guesser') {
      buttonRoleGuesser.disabled = true;
      buttonRoleSpymaster.disabled = false;
      toggleDifficulty.style.display = "none"
    } else {
      buttonRoleGuesser.disabled = false;
      buttonRoleSpymaster.disabled = true;
      toggleDifficulty.style.display = "block"
    }
    wipeBoard();
  }
})

socket.on('gameState', (data) =>{           // Response to gamestate update
  if (data.difficulty !== difficulty){  // Update the clients difficulty
    difficulty = data.difficulty
    wipeBoard();                        // Update the appearance of the tiles
  }
  mode = data.mode                      // Update the clients game mode
  consensus = data.consensus            // Update the clients consensus mode
  updateInfo(data.game, data.team)      // Update the games turn information
  updateTimerSlider(data.game, data.mode)          // Update the games timer slider
  updatePacks(data.game)                // Update the games pack information
  updatePlayerlist(data.players)        // Update the player list for the room

  let proposals = []
  for (let i in data.players){
    let guessProposal = data.players[i].guessProposal
    if (guessProposal !== null){
      proposals.push(guessProposal)
    }
  }

  // Update the board display
  updateBoard(data.game.board, proposals, data.game.over)
  updateLog(data.game.log)
})


// Utility Functions
////////////////////////////////////////////////////////////////////////////

// Wipe all of the descriptor tile classes from each tile
function wipeBoard(){
  for (let x = 0; x < 5; x++){
    let row = document.getElementById('row-' + (x+1))
    for (let y = 0; y < 5; y++){
      let button = row.children[y]
      button.className = 'tile'
    }
  }
}

// Update the game info displayed to the client
function updateInfo(game, team){
  scoreBlue.innerHTML = game.blue                         // Update the blue tiles left
  scoreRed.innerHTML = game.red                           // Update the red tiles left
  turnMessage.innerHTML = game.turn + "'s turn"           // Update the turn msg
  turnMessage.className = game.turn                       // Change color of turn msg
  if (game.over){                                         // Display winner
    turnMessage.innerHTML = game.winner + " wins!"
    turnMessage.className = game.winner
  }
  if (team !== game.turn) endTurn.disabled = true         // Disable end turn button for opposite team
  else endTurn.disabled = false
  if (playerRole === 'spymaster') {
    endTurn.disabled = true // Disable end turn button for spymasters
  }
  clueEntryDiv.style.display = playerRole === 'spymaster' && game.clue === null && team === game.turn ? '' : 'none'
  if (game.over || game.clue === null){
    clueDisplay.innerText = '___'
  }
  else {
    clueDisplay.innerText = game.clue.word + " (" + (game.clue.count === 'unlimited' ? '∞' : game.clue.count) + ")"
  }
}

// Update the clients timer slider
function updateTimerSlider(game, mode){
  let minutes = (game.timerAmount - 1) / 60
  timerSlider.value = minutes
  timerSliderLabel.innerHTML = "Timer Length : " + timerSlider.value + "min"

  // If the mode is not timed, dont show the slider
  if (mode === 'casual'){
    timerSlider.style.display = 'none'
    timerSliderLabel.style.display = 'none'
  } else {
    timerSlider.style.display = 'block'
    timerSliderLabel.style.display = 'block'
  }
}

// Update the pack toggle buttons
function updatePacks(game){
  if (game.base) buttonBasecards.className = 'enabled'
  else buttonBasecards.className = ''
  if (game.duet) buttonDuetcards.className = 'enabled'
  else buttonDuetcards.className = ''
  if (game.undercover) buttonUndercovercards.className = 'enabled'
  else buttonUndercovercards.className = ''
   if (game.custom) buttonCustomcards.className = 'enabled'
  else buttonCustomcards.className = ''
  if (game.nsfw) buttonNsfwcards.className = 'enabled'
  else buttonNsfwcards.className = ''  
  document.getElementById('word-pool').innerHTML = "Word Pool: " + game.words.length
}

// Update the board
function updateBoard(board, proposals, gameOver){
  // Add description classes to each tile depending on the tiles color
  for (let x = 0; x < 5; x++){
    let row = document.getElementById('row-' + (x+1))
    for (let y = 0; y < 5; y++){
      let button = row.children[y]
      button.innerHTML = board[x][y].word
      button.className = "tile"
      if (board[x][y].type === 'red') button.className += " r"    // Red tile
      if (board[x][y].type === 'blue') button.className += " b"   // Blue tile
      if (board[x][y].type === 'neutral') button.className += " n"// Neutral tile
      if (board[x][y].type === 'death') button.className += " d"  // Death tile
      if (board[x][y].flipped) button.className += " flipped"     // Flipped tile
      if (proposals.includes(board[x][y].word)) button.className += " proposed" // proposed guess
      if (playerRole === 'spymaster' || gameOver) button.className += " s"    // Flag all tiles if the client is a spy master
      if (difficulty === 'hard') button.className += " h"         // Flag all tiles if game is in hard mode
    }
  }
  // Show the proper toggle options for the game difficulty
  if (difficulty === 'normal') {
    buttonDifficultyNormal.disabled = true;
    buttonDifficultyHard.disabled = false;
  } else {
    buttonDifficultyNormal.disabled = false;
    buttonDifficultyHard.disabled = true;
  }
  // Show the proper toggle options for the game mode
  if (mode === 'casual') {
    buttonModeCasual.disabled = true;
    buttonModeTimed.disabled = false;
    timer.innerHTML = ""
  } else {
    buttonModeCasual.disabled = false;
    buttonModeTimed.disabled = true;
  }
  // Show the proper toggle options for the game consensus mode
  if (consensus === 'single') {
    buttonConsensusSingle.disabled = true;
    buttonConsensusConsensus.disabled = false;
  } else {
    buttonConsensusSingle.disabled = false;
    buttonConsensusConsensus.disabled = true;
  }
}

// Update the player list
function updatePlayerlist(players){
  undefinedList.innerHTML = ''
  redTeam.innerHTML = ''
  blueTeam.innerHTML = ''
  for (let i in players){
    // Create a li element for each player
    let li = document.createElement('li');
    li.innerText = players[i].nickname
    // If the player is a spymaster, put brackets around their name
    if (players[i].role === 'spymaster') li.innerText = "[" + players[i].nickname + "]"
    else if (players[i].guessProposal !== null){
      let guessProposal = document.createElement('span')
      guessProposal.classList.add('guess-proposal')
      guessProposal.innerText = players[i].guessProposal
      li.appendChild(guessProposal)
    }
    // Add the player to their teams ul
    if (players[i].team === 'undecided'){
      undefinedList.appendChild(li)
    } else if (players[i].team === 'red'){
      redTeam.appendChild(li)
    } else if (players[i].team === 'blue'){
      blueTeam.appendChild(li)
    }
  }
}

function updateLog(log){
  logDiv.innerHTML = ''
  log.forEach(logEntry => {
    let logSpan = document.createElement('span')
    logSpan.className = logEntry.event + " " + logEntry.team
    if (logEntry.event === 'flipTile'){
      logSpan.innerText = (logEntry.team + " team flipped " + logEntry.word
                           + " (" + logEntry.type + ")"
                           + (logEntry.type === 'death' ? " ending the game"
                              : logEntry.endedTurn ? " ending their turn"
                              : ""))
    }
    else if (logEntry.event === 'switchTurn'){
      logSpan.innerText = "Switched to " + logEntry.team + " team's turn"
    }
    else if (logEntry.event === 'declareClue'){
      logSpan.innerText = (logEntry.team + ' team was given the clue "'
                           + logEntry.clue.word + '" ('
                           + (logEntry.clue.count === 'unlimited'
                              ? '∞'
                              : logEntry.clue.count)
                           + ')')
    }
    else if (logEntry.event === 'endTurn'){
      logSpan.innerText = logEntry.team + " team ended their turn"
    }
    logDiv.prepend(logSpan)
  })
}

function extractFromFragment(fragment, toExtract) {
  let vars = fragment.substring(1).split('&');
  for (let i = 0; i < vars.length; i++) {
    var pair = vars[i].split('=');
    if (pair.length !== 2) {
      continue;
    }
    if (decodeURIComponent(pair[0]) === toExtract) {
        return decodeURIComponent(pair[1]);
    }
  }
  return '';
}

function updateFragment() {
  let room = joinRoom.value;
  let password = joinPassword.value;
  let fragment =  'room=' + encodeURIComponent(room) + '&password=' + encodeURIComponent(password);
  window.location.hash = fragment;
}

File: codenames.go
package main

import (
	"fmt"
	"sync"

	"github.com/sirupsen/logrus"
)

type CodeNames struct {
	sync.RWMutex
	PlayerRooms map[string]*Room
	NameRooms   map[string]*Room
}

func NewCodeNames() *CodeNames {
	return &CodeNames{
		PlayerRooms: map[string]*Room{},
		NameRooms:   map[string]*Room{},
	}
}

func (cn *CodeNames) PlayerRoom(playerID string) *Room {
	cn.RLock()
	defer cn.RUnlock()
	room := cn.PlayerRooms[playerID]
	if room == nil {
		return nil
	}
	return room
}

func (cn *CodeNames) PlayerRoomName(playerID string) string {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		return ""
	}
	return room.Name
}

func (cn *CodeNames) RoomName(roomName string) *Room {
	cn.RLock()
	defer cn.RUnlock()
	room, ok := cn.NameRooms[roomName]
	if !ok {
		return nil
	}
	return room
}

func (cn *CodeNames) SetNameRoom(roomName string, room *Room) {
	cn.Lock()
	cn.NameRooms[roomName] = room
	cn.Unlock()

}

func (cn *CodeNames) CleanRoom(room *Room) {
	cn.Lock()
	if len(room.Players) == 0 {
		delete(cn.NameRooms, room.Name)
	}
	cn.Unlock()
}

func (cn *CodeNames) DeletePlayer(playerID string) {
	cn.Lock()
	delete(cn.PlayerRooms, playerID)
	cn.Unlock()
}

func (cn *CodeNames) CreateRoom(playerID, nick, room, password string) (string, bool) {
	if oldRoom := cn.PlayerRoom(playerID); oldRoom != nil {
		oldRoom.Leave(playerID)
	}

	if r := cn.RoomName(room); r != nil {
		log.WithFields(logrus.Fields{
			"PlayerID": playerID,
			"Nick":     nick,
			"Password": password,
			"RoomName": room,
		}).Info("player tried to create room but already exists")
		return fmt.Sprintf("room %s already exists.", room), false
	}
	cn.SetNameRoom(room, NewRoom(room, password))
	return cn.JoinRoom(playerID, room, nick, password)
}

func (cn *CodeNames) JoinRoom(playerID, roomname, nick, password string) (string, bool) {
	if len(nick) == 0 {
		return "invalid nickname", false
	}
	if len(password) == 0 {
		return "invalid password", false
	}

	room := cn.RoomName(roomname)
	if room == nil {
		log.WithFields(logrus.Fields{
			"PlayerID": playerID,
			"Nick":     nick,
			"Password": password,
			"RoomName": room,
		}).Info("player tried to join a room but the room does not exist")

		return fmt.Sprintf("could not find room: %s", room.Name), false
	}

	if room.Password != password {
		return "invalid password", false
	}
	ok := room.Join(playerID, nick)
	if !ok {
		return "unable to join room", false
	}

	cn.Lock()
	cn.PlayerRooms[playerID] = room
	cn.Unlock()

	log.WithFields(logrus.Fields{
		"PlayerID":   playerID,
		"PlayerNick": nick,
		"RoomName":   room.Name,
	}).Info("player joined room")

	return "joined the room", true

}

func (cn *CodeNames) LeaveRoom(playerID string) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.WithFields(logrus.Fields{
			"PlayerID": playerID,
		}).Warn("player tried to leave the roo but they are not in any room")
		return false
	}

	ok := room.Leave(playerID)
	if !ok {
		log.WithFields(logrus.Fields{
			"PlayerID": playerID,
			"RoomName": room.Name,
		}).Warn("player was unable to leave the room")

		return false
	}

	cn.DeletePlayer(playerID)

	// room gets deleted when all players leave
	// todo(voldy): fix this, maybe?
	cn.CleanRoom(room)

	return true
}

func (cn *CodeNames) JoinTeam(playerID, team string) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.WithFields(logrus.Fields{
			"PlayerID": playerID,
			"RoomName": room.Name,
			"Team":     team,
		}).Info("player tried to join team but they are not in any room")
		return false
	}
	room.ChangeTeam(playerID, team)
	return true
}

func (cn *CodeNames) RandomizeTeams(playerID string) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.WithFields(logrus.Fields{
			"PlayerID": playerID,
		}).Warn("player tried to randomize teams but they are not in any room")
		return false
	}
	room.RandomizeTeams(playerID)
	return true
}

func (cn *CodeNames) NewGame(playerID string) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.WithFields(logrus.Fields{
			"PlayerID": playerID,
		}).Warn("player tried to start a new game but they are not in any room")
		return false
	}

	room.NewGame()

	return true
}

func (cn *CodeNames) SwitchRole(playerID, role string) (string, bool) {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.WithFields(logrus.Fields{
			"PlayerID": playerID,
			"Role":     role,
		}).Warn("player tried to switch role but they are not in any room")
		return "you can't switch roles when you are not part of a room", false
	}

	return room.SwitchRole(playerID, role)
}

func (cn *CodeNames) SwitchDifficulty(playerID, difficulty string) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.Printf("player %s tried to switch difficulty to %s but they are not in a room", playerID, difficulty)
		return false
	}

	room.ChangeDifficulty(playerID, difficulty)

	return true
}

func (cn *CodeNames) SwitchMode(playerID, mode string) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.WithFields(logrus.Fields{
			"PlayerID": playerID,
			"Mode":     mode,
		}).Info("switch mode failed, player was not found in a room")
		return false
	}

	room.SwitchMode(playerID, mode)
	return true
}

func (cn *CodeNames) SwitchConsensus(playerID, roomname, consensus string) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.Printf("player %s tried to switch consensus to %s but they are not in a room", playerID, consensus)
		return false
	}

	room.SwitchConsensus(playerID, consensus)
	return true
}

func (cn *CodeNames) EndTurn(playerID string) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.Printf("player %s tried end turn but they are not in a room", playerID)
		return false
	}

	room.EndTurn(playerID)
	return true
}

func (cn *CodeNames) ClickTile(playerID string, i, j int) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.Printf("player %s tried click a tile but they are not in a room", playerID)
		return false
	}

	room.SelectTile(playerID, i, j)
	return true
}

func (cn *CodeNames) DeclareClue(playerID, word string, count int) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.Printf("player %s tried to declare a clue %s for %d but they are not in a room", playerID, word, count)
		return false
	}

	room.DeclareClue(playerID, word, count)
	return true
}

func (cn *CodeNames) ChangeCards(playerID, pack string) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.Printf("player %s tried to change pack to %s but they are not in a room", playerID, pack)
		return false
	}

	room.ChangeCards(playerID, pack)
	return true
}

func (cn *CodeNames) UpdateTimeSlider(playerID string, value float64) bool {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.Printf("player %s tried to update timer to value %d but they are not in a room", playerID, value)
		return false
	}

	room.ChangeTimer(playerID, value)
	return true
}

func (cn *CodeNames) PlayerGameState(playerID string) *gameState {
	room := cn.PlayerRoom(playerID)
	if room == nil {
		log.WithField("PlayerID", playerID).Info("room not found for player to retrieve game state")
		return nil
	}

	p, ok := room.Player(playerID)
	if !ok {
		logrus.WithFields(logrus.Fields{
			"PlayerID": playerID,
			"RoomName": room.Name,
		}).Warn("player was mapped to a room but they were not in the room")

		return nil
	}
	return &gameState{
		Room:       room.Name,
		Game:       room.Game,
		Difficulty: room.Difficulty,
		Consensus:  room.Consesus,
		Mode:       room.Mode,
		Players:    room.Players,
		Team:       p.Team,
	}
}

func (cn *CodeNames) RoomNameGameState(roomName string) *gameState {
	r := cn.RoomName(roomName)
	if r == nil {
		log.WithFields(logrus.Fields{
			"RoomName": roomName,
		}).Warn("room not found for looking up state")
		return nil
	}
	return cn.roomGameState(r)
}

func (cn *CodeNames) roomGameState(room *Room) *gameState {
	return &gameState{
		Room:       room.Name,
		Game:       room.Game,
		Difficulty: room.Difficulty,
		Consensus:  room.Consesus,
		Mode:       room.Mode,
		Players:    room.Players,
	}
}

func (cn *CodeNames) Players() int {
	return len(cn.PlayerRooms)
}

func (cn *CodeNames) Rooms() int {
	return len(cn.NameRooms)
}

type gameState struct {
	Room       string             `json:"room"`
	Players    map[string]*Player `json:"players"`
	Game       *Game              `json:"game"`
	Difficulty string             `json:"difficulty"`
	Mode       string             `json:"mode"`
	Consensus  string             `json:"consensus"`
	Team       string             `json:"team"`
}
